### ISSUE TITLE
"wrong curve" error during mutual authentication since upgrade from 1.1.0l to 1.1.1n

### ISSUE THREAD

**Issue by christopheronco** [openssl/openssl#18280]
My apache server has recently been upgraded. Openssl version changed from _**OpenSSL 1.1.0l  10 Sep 2019**_ to _**OpenSSL 1.1.1n  15 Mar 2022**_.

Since the upgrade, connections from client where mutual authentication is required fail. I have reproduced the issue using "openssl s_client" and "openssl s_server".

Whit openssl 1.1.1n:
 - client side:
```
cro@woa4test:~/woa07746$ openssl s_client -cert ztp.pem -key kerlink_private.key -connect 127.0.0.1:4433
CONNECTED(00000003)
...
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: 
    Session-ID-ctx: 
    Master-Key: 02ADECCE765D7993AA5FEB35484B295B62B06F7DE75DF9CED4DE825D66BD92D5031989D14833608734679877C7DEE6DB
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1652185495
    Timeout   : 7200 (sec)
    Verify return code: 21 (unable to verify the first certificate)
    Extended master secret: yes
---
```

 - server side:
```
cro@woa4test:~/woa07746$ openssl version
OpenSSL 1.1.1n  15 Mar 2022
cro@woa4test:~/woa07746$ openssl s_server -accept *:4433 -key wanesy.com.key -cert wanesy.com.crt -Verify 3 -verify_return_error -verifyCAfile KerlinkRootCA.crt -tls1_2
verify depth is 3, must return a certificate
Using default temp DH parameters
ACCEPT
depth=2 CN = KerlinkRootCA
verify return:1
depth=1 CN = WOACA
verify return:1
depth=0 CN = 7276FF002E06029C
verify return:1
ERROR
139803539948736:error:1414D17A:SSL routines:tls12_check_peer_sigalg:wrong curve:../ssl/t1_lib.c:1095:
shutting down SSL
CONNECTION CLOSED
```
With openssl 1.1.0l:
 - client side:
```
cro@woa:~/woa07746$ openssl s_client -cert ztp.pem -key kerlink_private.key -connect 127.0.0.1:4433
CONNECTED(00000003)
...
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: 0457C4A216C85EE659B698D460D2CF6E01D6B6AC6862F5F16BEC19F8FFA2570D
    Session-ID-ctx: 
    Master-Key: A42FAE397025169CD11EE801DB5BADBDCADBBFA16394A381A7BFADA1150C3489AFDB27FD8345C0A1AEDEA6961A92B442
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 7200 (seconds)
    TLS session ticket:
    0000 - fa b9 39 8d dc f6 80 f3-c5 1b 28 da 63 f9 33 15   ..9.......(.c.3.
    0010 - 0a 3a 86 67 e1 5a e4 b9-f0 f5 ae 73 17 6b 11 af   .:.g.Z.....s.k..
    0020 - ed bc ec 29 36 01 5f fb-86 ee 67 d2 6d a4 8c ae   ...)6._...g.m...
    0030 - 7d 32 50 cb fd 25 2d 1e-c3 88 f7 d8 57 2a 53 d0   }2P..%-.....W*S.
    0040 - 3c b6 cd d1 32 bd dd 94-e8 9d 69 39 9f 7a 19 07   <...2.....i9.z..
    0050 - 22 fb 28 54 74 77 f8 70-9f 5c 43 8d 50 af b5 74   ".(Ttw.p.\C.P..t
    0060 - 00 c1 07 da 6a ca 95 55-e7 29 93 af 85 d3 d3 37   ....j..U.).....7
    0070 - 90 ac 18 ae 0e a2 75 55-ef 28 94 bf 3c 29 c7 5f   ......uU.(..<)._
    0080 - bb 2f 9e ec 35 60 a5 a8-8a 70 95 aa fc b1 ed 86   ./..5`...p......
    0090 - 7a 28 b8 6c 4e 18 7d 3f-d2 d4 c2 71 26 9a 0e cf   z(.lN.}?...q&...
    00a0 - fc aa e2 21 30 2f b7 2b-7b 92 a4 23 c6 00 a5 62   ...!0/.+{..#...b
    00b0 - 07 a4 e6 8b 8c 87 71 84-c6 1a 07 5a 4c cc bf 80   ......q....ZL...
    00c0 - 77 0c ea a2 af 37 f9 c7-89 e8 53 f7 70 cb 30 5a   w....7....S.p.0Z
    00d0 - dd ba 96 36 f9 9e aa 38-24 e6 c8 4c f6 68 de 83   ...6...8$..L.h..
    00e0 - 42 a6 19 83 75 14 fc 13-79 a6 e0 0d 0c 33 dd 43   B...u...y....3.C
    00f0 - 04 ab 06 0c d7 27 90 e6-13 a6 35 73 f7 ee 36 cb   .....'....5s..6.
    0100 - 74 f1 7a 12 f7 12 26 3e-c7 b3 9d 2d 08 a5 27 08   t.z...&>...-..'.
    0110 - b2 35 e9 92 bc c2 71 30-41 b1 57 ac ac 44 29 21   .5....q0A.W..D)!
    0120 - fe f1 0e 6b 73 09 d5 56-b6 2d d7 f6 c3 36 16 16   ...ks..V.-...6..
    0130 - 38 50 6b ad 59 3f 63 57-50 10 9b 78 ec 96 66 2f   8Pk.Y?cWP..x..f/
    0140 - 69 c2 18 98 c2 a7 b6 c4-97 f1 e0 ad 99 c7 5c ee   i.............\.
    0150 - e1 32 15 65 93 02 55 a6-9d 72 0c 55 e5 02 5a cf   .2.e..U..r.U..Z.
    0160 - 8f fb 7f 2c 1d da 42 60-63 9b dc 7a 9e a6 5e 7b   ...,..B`c..z..^{
    0170 - ac fa 03 3c be 12 eb 5b-95 76 e4 85 42 a8 70 c7   ...<...[.v..B.p.
    0180 - 44 60 ad 9d 7e a0 45 8a-40 ce 8d 71 ce 2d 66 32   D`..~.E.@..q.-f2
    0190 - 56 2a c2 49 25 d0 fe 41-99 b8 50 ee 9b 0b 5d 40   V*.I%..A..P...]@
    01a0 - ad a7 03 35 94 2a de 64-cd 2a 25 13 58 b3 3a 4e   ...5.*.d.*%.X.:N
    01b0 - 80 2e 35 77 d9 1a 6d 21-0d 75 5f 10 c0 13 9a 70   ..5w..m!.u_....p
    01c0 - 89 04 28 ce c2 81 9a 8d-f3 a1 03 a2 16 42 15 a0   ..(..........B..
    01d0 - 66 f9 83 62 c0 02 0c 55-79 3a 33 dd c1 74 e6 08   f..b...Uy:3..t..
    01e0 - 7c 1e 66 95 24 b7 5c bb-b3 5a 4f 57 4f 8b 05 11   |.f.$.\..ZOWO...
    01f0 - 21 d1 5e 94 4c 6c bf f9-ad dd 36 5b 32 46 8e 67   !.^.Ll....6[2F.g
    0200 - 75 29 eb 6b 3a 1b 03 3e-ce 45 a0 14 b8 ed c3 a2   u).k:..>.E......
    0210 - 5f 56 9f 6c 5d 1c a8 ec-09 3a aa 73 4d 28 ad 01   _V.l]....:.sM(..
    0220 - 2a 5e 25 f0 81 3c bd 31-5c e7 14 df 48 b0 3a 5f   *^%..<.1\...H.:_
    0230 - b4 1d c4 60 c4 11 4d 8a-02 dd 0a 0b 0e 07 d9 65   ...`..M........e
    0240 - cd 02 bc 70 2c ae 3b 23-cf 8f 2d 3d 9b 32 6b 26   ...p,.;#..-=.2k&
    0250 - d5 b7 21 15 fb b4 0c f4-a7 60 3b bd 29 e7 59 9d   ..!......`;.).Y.
    0260 - a3 45 7f ae fe 72 c3 46-96 0c 7c 04 19 f6 40 a3   .E...r.F..|...@.
    0270 - bf ed d6 eb b9 3e 21 ef-ee 54 40 d6 72 a0 aa d3   .....>!..T@.r...
    0280 - d6 be 8f 01 40 77 61 f5-fc 43 4e ff 14 32 b7 5a   ....@wa..CN..2.Z
    0290 - 09 02 46 06 dc 9e 14 3a-a3 27 94 36 19 ba b8 b9   ..F....:.'.6....
    02a0 - 09 be da 19 45 a3 1d 67-80 f9 44 df ea 37 87 06   ....E..g..D..7..
    02b0 - 1b 4a c7 3b b2 9e 08 d8-c9 1e 5f ee e8 0d 02 2f   .J.;......_..../
    02c0 - 22 f3 22 dd 3c 7d d8 72-1d c5 2a 88 62 c0 a4 8f   ".".<}.r..*.b...
    02d0 - 4e ca b7 c3 12 c9 a9 e5-54 10 5a 10 12 69 f9 5a   N.......T.Z..i.Z
    02e0 - 18 55 9b 2e 18 15 8f 9b-c6 1c f8 2c 12 ff dd 8b   .U.........,....
    02f0 - 21 9d 26 ab ae db 2a 76-3b 49 60 c1 d0 7a 21 e4   !.&...*v;I`..z!.

    Start Time: 1652185854
    Timeout   : 7200 (sec)
    Verify return code: 21 (unable to verify the first certificate)
    Extended master secret: yes
---
read:errno=0
```
 - server side:
```
cro@woa:~/woa07746$ openssl version
OpenSSL 1.1.0l  10 Sep 2019
cro@woa:~/woa07746$ openssl s_server -accept *:4433 -key wanesy.com.key -cert wanesy.com.crt -Verify 3 -verify_return_error -verifyCAfile KerlinkRootCA.crt -tls1_2
verify depth is 3, must return a certificate
Using default temp DH parameters
ACCEPT
depth=2 CN = KerlinkRootCA
verify return:1
depth=1 CN = WOACA
verify return:1
depth=0 CN = 7276FF002E06029C
verify return:1
-----BEGIN SSL SESSION PARAMETERS-----
MIICtgIBAQICAwMEAsAwBAAEMKQvrjlwJRac0R7oAdtbrb3K27+hY5Sjgae/raEV
DDSJr9sn/YNFwKGu3qaWGpK0QqEGAgRielr+ogQCAhwgo4ICWDCCAlQwggE8oAMC
AQICCD7ZyRxxZZyXMA0GCSqGSIb3DQEBDQUAMBAxDjAMBgNVBAMMBVdPQUNBMB4X
DTIyMDMyOTA4MzMwMVoXDTI0MDMyODA4MzMwMVowGzEZMBcGA1UEAwwQNzI3NkZG
MDAyRTA2MDI5QzBWMBAGByqGSM49AgEGBSuBBAAKA0IABCAJGoEL0s1gsBYHmgsr
5IVQO4l3hnS8OsHFQqnAj+7AYPuyV5kmwRX60amr4D0iKdBR/AYfr2GFIEZ1qZp4
MoWjdTBzMB0GA1UdDgQWBBTtp4ShNEwCMb5sH/o1lIe0FM78MTAMBgNVHRMBAf8E
AjAAMB8GA1UdIwQYMBaAFH9iFB7M+my4GpxoEFxEtZQ6ZZmXMA4GA1UdDwEB/wQE
AwIDiDATBgNVHSUEDDAKBggrBgEFBQcDAjANBgkqhkiG9w0BAQ0FAAOCAQEAg2Wg
9FJ8CfoE4B94qB/F5MyhORAKRr5hgjHEJgX5Mqtk0XFcetV1O+k4cwyUCGqLEU9z
RjORmZDlBr6Q4vnCqdOg4rQJNtNyndgAmYo1uUTkDZusI4WWmnq0/QbomOVvr1tG
E6gju0wyNUWEjYIi0FyTE6muNrlY7KVOF//1k//ZYZqa0pyuMCFbqwIviJi2D6e5
v8PmHM/IXRUNsWY5mG2iyz/Gh8E6UgodfXm7bxUlzY6k77R1FP/BYlWewlY0s/Ye
5dzGJRXMwMVnQLS4PDufdFwpZCF2Iq9khK3SX9MIuP2GL/Th3nGF7lkTN2QOaMev
xvFkqLV4xui2qCVIm6QGBAQBAAAArQMCAQE=
-----END SSL SESSION PARAMETERS-----
Client certificate
-----BEGIN CERTIFICATE-----
MIICVDCCATygAwIBAgIIPtnJHHFlnJcwDQYJKoZIhvcNAQENBQAwEDEOMAwGA1UE
AwwFV09BQ0EwHhcNMjIwMzI5MDgzMzAxWhcNMjQwMzI4MDgzMzAxWjAbMRkwFwYD
VQQDDBA3Mjc2RkYwMDJFMDYwMjlDMFYwEAYHKoZIzj0CAQYFK4EEAAoDQgAEIAka
gQvSzWCwFgeaCyvkhVA7iXeGdLw6wcVCqcCP7sBg+7JXmSbBFfrRqavgPSIp0FH8
Bh+vYYUgRnWpmngyhaN1MHMwHQYDVR0OBBYEFO2nhKE0TAIxvmwf+jWUh7QUzvwx
MAwGA1UdEwEB/wQCMAAwHwYDVR0jBBgwFoAUf2IUHsz6bLganGgQXES1lDplmZcw
DgYDVR0PAQH/BAQDAgOIMBMGA1UdJQQMMAoGCCsGAQUFBwMCMA0GCSqGSIb3DQEB
DQUAA4IBAQCDZaD0UnwJ+gTgH3ioH8XkzKE5EApGvmGCMcQmBfkyq2TRcVx61XU7
6ThzDJQIaosRT3NGM5GZkOUGvpDi+cKp06DitAk203Kd2ACZijW5ROQNm6wjhZaa
erT9BuiY5W+vW0YTqCO7TDI1RYSNgiLQXJMTqa42uVjspU4X//WT/9lhmprSnK4w
IVurAi+ImLYPp7m/w+Ycz8hdFQ2xZjmYbaLLP8aHwTpSCh19ebtvFSXNjqTvtHUU
/8FiVZ7CVjSz9h7l3MYlFczAxWdAtLg8O590XClkIXYir2SErdJf0wi4/YYv9OHe
cYXuWRM3ZA5ox6/G8WSotXjG6LaoJUib
-----END CERTIFICATE-----
subject=/CN=7276FF002E06029C
issuer=/CN=WOACA
Shared ciphers:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:DHE-RSA-AES256-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:DHE-RSA-AES128-SHA:AES256-GCM-SHA384:AES128-GCM-SHA256:AES256-SHA256:AES128-SHA256:AES256-SHA:AES128-SHA
Signature Algorithms: RSA+SHA512:DSA+SHA512:ECDSA+SHA512:RSA+SHA384:DSA+SHA384:ECDSA+SHA384:RSA+SHA256:DSA+SHA256:ECDSA+SHA256:RSA+SHA224:DSA+SHA224:ECDSA+SHA224:RSA+SHA1:DSA+SHA1:ECDSA+SHA1
Shared Signature Algorithms: RSA+SHA512:DSA+SHA512:ECDSA+SHA512:RSA+SHA384:DSA+SHA384:ECDSA+SHA384:RSA+SHA256:DSA+SHA256:ECDSA+SHA256:RSA+SHA224:DSA+SHA224:ECDSA+SHA224:RSA+SHA1:DSA+SHA1:ECDSA+SHA1
Peer signing digest: SHA512
Supported Elliptic Curve Point Formats: uncompressed:ansiX962_compressed_prime:ansiX962_compressed_char2
Supported Elliptic Curves: X25519:P-256:P-521:P-384
Shared Elliptic curves: X25519:P-256:P-521:P-384
CIPHER is ECDHE-RSA-AES256-GCM-SHA384
Secure Renegotiation IS supported
```

I found that the problem is related to the private key of client certificate. It has been generated like that:
```
openssl ecparam -genkey -name secp256k1 -noout -out kerlink_private.key
```

If I use prime256v1 instead of secp256k1 to generate client private key, like that:
```
openssl ecparam -genkey -name prime256v1 -noout -out private_prime256v1.key
```
I have no problem to connect with mutual authentication even with opensssl 1.1.1n. But it is really difficult for me to change clients.
So:
 - is this normal?
 - is there an option on server side to authorize usage of secp256k1 curve in client certificates  in openssl 1.1.1n version?

**Comment by t8m** [comment]
You might also need to limit the TLS version on the client side to TLS-1.2.

**Comment by mattcaswell** [comment]
>  I will still need to change something on client side which is quite difficult for me

I thought you were trying to upgrade the server only? If you don't change the OpenSSL version on the client then you shouldn't need to make any configuration changes there either.

**Comment by christopheronco** [comment]
Yes I am trying to upgrade the server only. And yes I would like to make no change on client side. I have many clients and they use different versions of Openssl. I have seen: 
 - OpenSSL 1.1.1k  25 Mar 2021
 - OpenSSL 1.0.2n  7 Dec 2017

But I can miss some other versions.
I have tried:
 - client OpenSSL 1.1.1k and server OpenSSL 1.1.0l: OK
 - client OpenSSL 1.0.2n and server OpenSSL 1.1.0l: OK
 - client OpenSSL 1.1.1k and server OpenSSL 1.1.1n (with groups set to X25519:P-256:P-521:P-384:secp256k1): KO
 - client OpenSSL 1.0.2n and server OpenSSL 1.1.1n (with groups set to X25519:P-256:P-521:P-384:secp256k1): OK

So I assume I still need to configure groups on client side if client is using OpenSSL 1.1.1k. I have not tried because I don't know how to do that. These clients use ssl module from python 3.8.2 and I don't know how to set this option in this context.

I made a wireshark capture and with a client using OpenSSL 1.1.1k, supported groups in client hello does not contain secp256k1. With a client using OpenSSL 1.0.2n, supported Groups in client hello contains secp256k1. 
Is this the root cause or am I not looking at the right place?
If this is the root cause, why a client with OpenSSL 1.1.1k is able to connect if server is in OpenSSL 1.1.0l?

**Comment by mattcaswell** [comment]
I've dug into this a bit deeper. OpenSSL 1.1.1 is stricter than 1.1.0 is with regards to the supported groups extension.

In 1.1.0 an OpenSSL server would not check that a group used in a certificate sent by a client was in the list of supported groups originally sent by the client. In 1.1.1 we do enforce that check. So in this case, a 1.1.1 client will send its default list of supported groups ("X25519:P-256:P-521:P-384"), and then send a certificate that uses a different group ("secp256k1"). A 1.1.1 server will barf on this.

The change in behaviour was introduced by commit 4a1b4280199 in #4412. The code has moved around quite a bit between 1.1.0 and 1.1.1, but in 1.1.1 the new check looks like this:

https://github.com/openssl/openssl/blob/6ef91d8153e04a2302bff11b29caf7e888b62fe8/ssl/t1_lib.c#L1092-L1097

The 1.1.0 code looks like this:

https://github.com/openssl/openssl/blob/a3b54f0f5de1ad17889fd23aee7c230eefc300cd/ssl/t1_lib.c#L783-L787

Ignoring the irrelevant changes to the function that actually does the checking you can see that in the 1.1.1 code we perform the check regardless of whether we are a server or a client, but in 1.1.0 we only do the check if we are a client.

It is not clear from the commit description or the comments in #4412 whether this was an intentional change or not - at least there is no justification provided for it and no associated entry in CHANGES.

The question now is whether this should be considered a bug or not. The new 1.1.1 behaviour is not wrong, and is arguably more correct than the 1.1.0 version. OTOH it seems harmless to accept such a certificate and arguably we should maintain the behaviour from older versions - so in that sense it could be considered a regression.

The fact that an OpenSSL client will happily send a certificate using a group which we it did not declare in its supported groups extension is also dubious and could be treated as a bug - but this at least seems to be long standing behaviour and not a regression. Fixing it would either (potentially) break lots of clients (by suddenly disallowing them to send a certificate that they were previously allowed to send), or be quite complicated (because we have to inspect the certificate up-front, figure out what group it uses, and add it to the supported groups list automatically).

**Comment by christopheronco** [comment]
Thanks for the clear explanation. I still don't know how I will fix my problem on field but I am happy to participate in pointing it out.

**Comment by mattcaswell** [comment]
Since it is not clear whether we should treat this as a bug or not, I think this needs to go to OTC for discussion

**OTC Question: Do we consider it a bug that 1.1.1 verifies that the group in the client cert is in the client supported groups and rejects if not - when 1.1.0 did not enforce this check? This change of behaviour can break some setups**

**Comment by t8m** [comment]
**OTC: The server should be changed to accept the client certificate in TLS <= 1.2 if the curve is among the supported groups configured on the server side. We consider this is a bug fix.**

**Comment by t8m** [comment]
@mattcaswell Is this still relevant? In particular in regards to the OTC decision above.

**Comment by nhorman** [comment]
Marking as inactive, to be closed when 3.4 dev cycle ends, barring further input
